### PR TITLE
FLINK-3967 - Flink Sink for Rethink Db

### PR DESCRIPTION
**Pull request by mans2singh**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.
- [ ] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)
- [ ] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added
- [ ] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

### CODE REVIEW

File: flink-streaming-connectors/flink-connector-rethinkdb/src/main/java/org/apache/flink/streaming/connectors/rethinkdb/FlinkRethinkDbSink.java [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+		LOG.debug("Received value {}", value);
+		
+		Object json = serializationSchema.toJSON(value);
+		LOG.debug("Object/Json: {}/{}", value, json);
+		Insert insert = getRdbTable().insert(json).optArg(CONFLICT_OPT, conflict.toString());
+		HashMap<String,Object> result = runInsert(insert);
+		
+		LOG.debug("Object/Json/Result: {}/{}/{}", value, json, result);
+		
+		if ( (Long)result.get(RESULT_ERROR_KEY) != 0 ) {
```
Comment (zentol) [inline flink-streaming-connectors/flink-connector-rethinkdb/src/main/java/org/apache/flink/streaming/connectors/rethinkdb/FlinkRethinkDbSink.java]: this is a synchronous operation, correct? If so I'd be curious about a benchmark for this sink.

File: flink-streaming-connectors/flink-connector-rethinkdb/src/main/java/org/apache/flink/streaming/connectors/rethinkdb/FlinkRethinkDbSink.java [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+ * checks if there is an error entry in the result HashMap and throws a runtime exception if errors
+ * counts is not zero.  The exception message contains the results HashMap. 
+ * In case of multiple errors only the first error is noted in the result HashMap.
+ * </b>
+ * 
+ * @see {@link ConflictStrategy} for conflict resolution strategies
+ * 
+ * @param <OUT> a value that can be transformed into a {@link org.json.simple.JSONArray;} or {@link org.json.simple.JSONObject}
+ */
+public class FlinkRethinkDbSink<OUT> extends RichSinkFunction<OUT> implements Serializable{
```
Comment (zentol) [inline flink-streaming-connectors/flink-connector-rethinkdb/src/main/java/org/apache/flink/streaming/connectors/rethinkdb/FlinkRethinkDbSink.java]: This class should be named RethinkDBSink

**Comment by zentol** [comment]
All travis builds failed.

For Java 8 several RethinkDB tests failed.

For Java 7 we get an Unsupported Major/Minor version error. Does this Sink (or rather the driver) support Java 7?

**Comment by mans2singh** [comment]
Hey @zentol - Thanks for your feedback. 
-  I've renamed the sink as you had recommended.  
- Currently the driver from rethinkdb supports java 8 only.  There is one supporting java 6 but it no longer being maintained.
-  The current travis failures for jdk 7 are because of the version issue as you had mentioned and those for java 8 are due to failures in the yarn-tests module. 
- I've not done any performance tests.  If you have any suggestions for throughput numbers that might be valuable, or if there is any other sink that I can use for setting up baseline test, please let me know.

If you have any other advice/recommendation, please let me know.   

Thanks again

File: flink-streaming-connectors/flink-connector-rethinkdb/pom.xml [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+            <artifactId>flink-test-utils_2.10</artifactId>
+            <version>${project.version}</version>
+      			<type>test-jar</type>
+            <scope>test</scope>
+        </dependency>
+	<dependency>
+	    <groupId>com.rethinkdb</groupId>
+	    <artifactId>rethinkdb-driver</artifactId>
+	    <version>${rethinkdb.version}</version>
+        </dependency>
```
Comment (rmetzger) [inline flink-streaming-connectors/flink-connector-rethinkdb/pom.xml]: Whats the license of this driver?
Rethingdb seems to use the Gnu affero license (https://github.com/rethinkdb/rethinkdb/blob/next/COPYRIGHT), which is not compatible with the apache license http://www.apache.org/legal/resolved.html#category-x so we can not accept this contribution.

Comment (mans2singh) [inline flink-streaming-connectors/flink-connector-rethinkdb/pom.xml]: Hi @rmetzger 

Thanks for pointing the license issue.  I wanted to get your advice on two other options:

1. Can I contribute it as as other Flink contributions (under flink-contrib module) ?
2. There is another java-driver available for RethinkDB (https://github.com/npiv/rethink-java-driver), available under Apache License 2.0.  If that is acceptable for a contribution to Flink, I can see if I can migrate the sink to that driver.  

If you have any other recommendations, please let me know.

Thanks.

Comment (StephanEwen) [inline flink-streaming-connectors/flink-connector-rethinkdb/pom.xml]: Unfortunately, anything that ends up in the Flink git repository must be Apache License compatible, regardless of whether it is `flink-core` or `flink-contrib`.

Comment (mans2singh) [inline flink-streaming-connectors/flink-connector-rethinkdb/pom.xml]: Hi @StephanEwen - The 2nd option mentioned above (https://github.com/npiv/rethink-java-driver), is available under Apache License 2.0.  Is that acceptable for the Flink contribution ?
Thanks

**Comment by rmetzger** [comment]
Hi @mans2singh,
thanks a lot for this contribution. I wonder what's your motivation to implement a flink --> rethinkdb connector? I'm asking because the Flink community can not accept every contribution, in particular if its adding a large new component we need to maintain.
Only if there are enough users and developers willing support a feature, we can accept the contribution.
I'm a bit sorry to tell you this after you've implemented the code, on the other hand, our guidelines [1] clearly state that new features need to be discussed with the community first.

The good news is that we have a special section for community maintained modules on our website [2]. Therefore, I suggest that you put the code into a separate github repository and open a pull request to add a link to that repository.
The good thing about this approach is also, that we don't have to fix the license issue immediately.
Also, it doesn't mean that I'm against this contribution at all. If we see many users using the module from your github repository, and there are people in the community willing to maintain it, we can also look into merging it.

[1] http://flink.apache.org/contribute-code.html#before-you-start-coding
[2] http://flink.apache.org/community.html#third-party-packages

**Comment by mans2singh** [comment]
Hi @rmetzger 

Thanks for your advice/suggestions.  I will try to answer your questions below:

Regarding motivation for the connector - I started working with Flink late last year and found it had some unique features (like streaming framework based on events from the grounds up and flexibility of windowing options).  I am working through some real-time data flow use cases, where these capabilities can stream line our processing pipelines. The integration with RethinkDB came into play because from the dev perspective it is schema less, can ingest streams/batch of data, has map/reduce functionality.  From the ops/scaling perspective it can be scaled/re-sharded in real-time without downtime. It can also provide change streams for a table, or a document.  IMHO, Flink and Rethink complement each other for scalable, stream processing/analytics use cases.  So, I thought it might be good to contribute back to the open source community and therefore the PR.

Regarding guidelines for code-contribution - I did go through the document and I thought this PR would be in the same vein as the other streaming connectors (kafka, etc), and complementing them.  There was no new API, or change in interfaces, and the code base is pretty light weight because Flink/RethinkDB both make it easy to integrate. However, I did not realize that it would be considered to a new feature and required a design review but perhaps that was my oversight.

In any case, I really appreciate, the time you and your team took to review the PR and advice me.  Please let me know what's your recommendation on how I should proceed.

Thanks, Mans

**Comment by mans2singh** [comment]
Hey @rmetzger, @StephanEwen, Flink Folks:  

Just wanted to let you know that the rethinkdb java driver that I used for Flink integration is available under Apache License 2.0  (https://github.com/rethinkdb/rethinkdb/blob/next/drivers/COPYRIGHT).   

I've checked travis build and the rethink connector is passing but the last failure was in the flnk-yarn-tests module.

Please let me know if you have any advice/comments for me.

Thanks for your time.

**Comment by StephanEwen** [comment]
Hi @mans2singh !

This looks like a good piece of work, and ReThinkDB is an interesting technology!
An Apache-compatible licence is also good.

We are facing a bit of a maintenance issue right now, though, with a large number of new projects. There are so many possible DBs and connectors, and it is tough for the community to maintain them all. For that reason, I would suggest to stage such contributions of new and interesting connectors.

Could you put this into an standalone GitHub project for now? We would link this repository from the Apache Flink website and that way make it visible to the Flink community.
When we see that we get frequent requests for a ReThinkDB, we would like to move this to the core Flink code base.

Following this idea, we would end up with frequently used connectors in the core Flink repository, and a larger number of connectors in third party community repositories. I think that would help the stability of the project and all connectors.

Stephan

**Comment by rmetzger** [comment]
Hi @mans2singh !
There's good news regarding streaming connector contributions in Apache Flink: If you are interested in hosting the code as part of an Apache project, the Apache Bahir project has now a repository for hosing Flink connectors as well: https://github.com/apache/bahir-flink
If you would like to contribute the connector there, feel free to open a new pull request there.

Regards,
Robert

**Comment by mans2singh** [comment]
Hi @rmetzger - 
Thanks for the update.  
Can you please let me know the steps for contributing to bahir ?  Can I just fork it and add another project to the root folder ?  
Thanks

**Comment by rmetzger** [comment]
Hi @mans2singh,
I'm happy to hear that you are interested in contributing the code to Bahir.

You can just fork the bahir-flink repo and create a folder similar to the existing connectors.

**Pull request closed without merging**